# Read Data Demo

### Environment

If use MacBook and you don't have the requests library installed in your Python environment, run the following command line:

In [ ]:
pip install requests

### Demo

#### Get Database ID

1. import the sqlDataReading depot for data file:

In [4]:
import data.sqlDataFetch as sdf

2. Get database id by full name or year:

(a) By full name

In [5]:
print("Database id: ", sdf.get_id_by_name('APPROVED BUILDING PERMITS'))

Database id:  6ddcd912-32a0-43df-9908-63574f8c7e77


(b) By year number

In [6]:
print("Database id: ", sdf.get_id_by_year(2029))

Database id:  Invalid year, please provide a year between 2004 and 2024


In [7]:
database_id = sdf.get_id_by_year(2022)
print("Database id: ", database_id)

Database id:  4b99718b-d064-471b-9b24-517ae5effecc


#### Describe Database

To describe a database and see what keys it contains (e.g., column or field names)

In [5]:
sdf.describe_database(database_id)

Column Name          | Type      
--------------------------------
_id                  | int       
PID                  | text      
CM_ID                | text      
GIS_ID               | text      
ST_NUM               | text      
ST_NAME              | text      
UNIT_NUM             | text      
CITY                 | text      
ZIPCODE              | text      
BLDG_SEQ             | text      
NUM_BLDGS            | text      
LUC                  | text      
LU                   | text      
LU_DESC              | text      
BLDG_TYPE            | text      
OWN_OCC              | text      
OWNER                | text      
MAIL_ADDRESSEE       | text      
MAIL_ADDRESS         | text      
MAIL_CITY            | text      
MAIL_STATE           | text      
MAIL_ZIPCODE         | text      
RES_FLOOR            | text      
CD_FLOOR             | text      
RES_UNITS            | text      
COM_UNITS            | text      
RC_UNITS             | text      
LAND_SF        

### Fetch Data

#### Fetch DATA Without Condition and parameter

1. Using fetch_data method only with database id (note: It will only fetch up to 32000 rows)

In [6]:
df = sdf.fetch_data(database_id)

if df is not None:
  print(df)

          _id                                         _full_text         PID  \
0           1  '0':33,34,35,65,103 '0100001000':1,2 '02128':1...  0100001000   
1           2  '0':31,32,33,64,92,93 '0100002000':1,2 '02128'...  0100002000   
2      178594  '0':26,27,28,32,38,39,40,41,42,43,44 '0.00':35...  2205666000   
3           3  '0':32,33,34,65,97,98 '0100003000':1,2 '02128'...  0100003000   
4           4  '0':32,33,34,65,103,104 '0100004000':1,2 '0212...  0100004000   
...       ...                                                ...         ...   
31995   32068  '-22':8 '-402':27 '0':31,32,33,36,44,45,46,47,...  0306455422   
31996   31951  '-108':8,28 '0':35,36,39,100,103 '02118':10,31...  0306455184   
31997   31952  '-109':8,27 '0':34,35,38,97,100 '02118':10,30 ...  0306455186   
31998   31953  '-110':8 '0':39,40,43,97,100 '02118':10 '025.9...  0306455188   
31999   31954  '-101':29 '-111':7 '0':33,34,35,36,43,44,45,46...  0306455190   

            CM_ID      GIS_ID      ST_N

#### Fetch Data Without Condition

2. Using fetch_data method only with database id and parameters

In [9]:
df = sdf.fetch_data(database_id, _id=int, CITY=str, GROSS_AREA=float, GROSS_TAX='money', ZIPCODE=int, LIVING_AREA=float, BLDG_VALUE='value')
if df is not None:
  print(df)

          _id         CITY  GROSS_AREA  GROSS_TAX  ZIPCODE  LIVING_AREA  \
0           1  EAST BOSTON      3353.0    7827.07     2128       2202.0   
1           2  EAST BOSTON      3299.0    8103.42     2128       2307.0   
2      178594     BRIGHTON         NaN       0.00     2135          NaN   
3           3  EAST BOSTON      3392.0    7947.84     2128       2268.0   
4           4  EAST BOSTON      3108.0    7266.75     2128       2028.0   
...       ...          ...         ...        ...      ...          ...   
31995   32068       BOSTON         1.0     435.20     2118          1.0   
31996   31951       BOSTON      1686.0   16495.17     2118       1686.0   
31997   31952       BOSTON      1678.0    6481.21     2118       1678.0   
31998   31953       BOSTON      1678.0   18025.98     2118       1678.0   
31999   31954       BOSTON         NaN    2165.36     2118          NaN   

       BLDG_VALUE  
0        539900.0  
1        564500.0  
2             0.0  
3        549700.0  

#### Fetch Data with All You Need

3. Using fetch_data method only with database id condition and parameters

In [8]:
condition= "WHERE \"CITY\"::text = 'ALLSTON'"
df = sdf.fetch_data(database_id, condition, _id=int, CITY=str, GROSS_AREA=float, GROSS_TAX='money', ZIPCODE=int, LIVING_AREA=float, BLDG_VALUE='value')
if df is not None:
  print(df)

         _id     CITY  GROSS_AREA  GROSS_TAX  ZIPCODE  LIVING_AREA  BLDG_VALUE
0     162575  ALLSTON         NaN    5043.46     2134          NaN    201900.0
1     162576  ALLSTON         NaN    6324.94     2134          NaN    253200.0
2     162577  ALLSTON         NaN    3010.09     2134          NaN    120500.0
3     162578  ALLSTON         NaN    4668.76     2134          NaN    186900.0
4     165044  ALLSTON         NaN    5890.28     2134          NaN    235800.0
...      ...      ...         ...        ...      ...          ...         ...
4435  174488  ALLSTON      2847.8   10041.15     2134       1969.8    697600.0
4436  175412  ALLSTON       772.0    4222.53     2134        772.0    388100.0
4437  175413  ALLSTON       774.0    3794.94     2134        774.0    348800.0
4438  175510  ALLSTON      3904.0    9598.34     2134       2928.0    695300.0
4439  177868  ALLSTON      3260.6    8733.38     2134       2217.6    604100.0

[4440 rows x 7 columns]


Here is  the mapping from Python data types to SQL data types and functions:
```python
type_mapping = {
    str: "TEXT",
    int: "INT",
    float: "FLOAT",
    bool: "BOOLEAN",
    # Additional mappings based on the provided SQL examples
    'timestamp': lambda col: f"TO_CHAR(\"{col}\"::timestamp, 'YYYY-MM-DD')",
    'value': lambda col: f"CAST(REPLACE(\"{col}\", ',', '') AS INT) AS \"{col}\"",
    'money': lambda col: f"CAST(REPLACE(REPLACE(\"{col}\", '$', ''), ',', '') AS FLOAT) AS \"{col}\"",
}


This $type\_mapping$ is not all. If you need other type matchings, you can add them to the $get\_sql\_cast$ method.

### Bones
Here are some SQL query examples proved by BOS gov:

```SQL
SELECT
    TO_CHAR("issued_date"::timestamp, 'YYYY') AS YEAR,
    COUNT(*) PERMITS,
    COUNT(DISTINCT "parcel_id") PROPERTIES,
    SUM("declared_valuation"::decimal)::money TOTAL_VALUE
FROM "6ddcd912-32a0-43df-9908-63574f8c7e77" APPROVED_BUILDING_PERMITS
WHERE
    "zip" = '02128'
    AND "address" NOT ILIKE '%%Harborside%%'
    AND "address" NOT ILIKE '%%Logan%%'
    AND "address" NOT ILIKE '%%Terminal%%'
    AND "declared_valuation"::decimal < 500000
GROUP BY TO_CHAR("issued_date"::timestamp, 'YYYY')
ORDER BY TO_CHAR("issued_date"::timestamp, 'YYYY') ASC;

SELECT
    *
FROM "6ddcd912-32a0-43df-9908-63574f8c7e77" APPROVED_BUILDING_PERMITS
LIMIT 1;

SELECT
    TO_CHAR("Date"::timestamp, 'YYYY') AS YEAR,
    SUM("Connections"::decimal) AS TOTAL_CONNECTIONS
FROM "3fe9760f-2e4b-4992-8c26-e2e6df59cf3b" WICKED_FREE_WIFI_CONNECTIONS
WHERE
    "Neighborhood" = 'East Boston'
GROUP BY TO_CHAR("Date"::timestamp, 'YYYY')
ORDER BY TO_CHAR("Date"::timestamp, 'YYYY') ASC;

SELECT
 "YEAR",
 count(*) TOTAL,
 count(*) FILTER (WHERE "UCR_PART" = 'Part One') PART_ONE,
 count(*) FILTER (WHERE "UCR_PART" = 'Part Two') PART_TWO,
 count(*) FILTER (WHERE "UCR_PART" = 'Part Three') PART_THREE,
 count(*) FILTER (WHERE "UCR_PART" = 'Other') OTHER
FROM "12cb3883-56f5-47de-afa5-3b1cf61b257b" CRIME_INCIDENT_REPORTS
WHERE
    "DISTRICT" = 'A7'
GROUP BY "YEAR"
ORDER BY "YEAR" asc;

SELECT
    *
FROM "2968e2c0-d479-49ba-a884-4ef523ada3c0" REPORTS_311
LIMIT 10;

SELECT
    TO_CHAR("open_dt"::timestamp, 'YYYY') AS YEAR,
    COUNT(*) TOTAL_REPORTS,
    COUNT(*)/MAX(CASE WHEN EXTRACT(YEAR FROM CURRENT_DATE) = EXTRACT(YEAR FROM "open_dt"::timestamp) THEN CURRENT_DATE-(EXTRACT(YEAR FROM CURRENT_DATE)||'-01-01')::date ELSE 365 END) DAILY_AVERAGE
FROM "2968e2c0-d479-49ba-a884-4ef523ada3c0" REPORTS_311
WHERE
    "neighborhood_services_district" = '1'
GROUP BY TO_CHAR("open_dt"::timestamp, 'YYYY')
ORDER BY TO_CHAR("open_dt"::timestamp, 'YYYY') ASC;

SELECT
    point("Lat"::decimal,"Long"::decimal) <@ circle '((42.370024,-71.0355957),0.01)' within1km
FROM "12cb3883-56f5-47de-afa5-3b1cf61b257b" CRIME_INCIDENT_REPORTS
WHERE "DISTRICT"::text = 'A7' and "Location"::text != '(0.00000000, 0.00000000)'
LIMIT 100;

WITH RANKED AS (SELECT
    RANK() OVER (ORDER BY sum("declared_valuation"::decimal) DESC) RANK,
    "property_id"::text,
    MAX("declared_valuation"::decimal) AS "TOTAL_VALUE"
FROM "6ddcd912-32a0-43df-9908-63574f8c7e77" APPROVED_BUILDING_PERMITS
WHERE
    "zip"::text = '02128'
    AND "owner"::text not in (
        'MASSACHUSETTS PORT AUTHORITY',
        'CITY OF BOSTON',
        'BOSTON HOUSING AUTHORITY',
        'ROMAN CATH ARCH OF BOS',
        'BROOKE CHARTER SCHOOL',
        'BROOKE CHARTER SCHOOL '
        'CITY OF BOSTON ',
        'MASSACHUSETTS PORT AUTH',
        'MASS PORT AUTHORITY',
        'CITY  OF  BOSTON',
        'BROOKE CHARTER SCHOOL '
    )
    AND "occupancytype"::text != 'Comm'
GROUP BY "property_id"::text
ORDER BY "TOTAL_VALUE" desc)
SELECT
    r.RANK,a.*
FROM "6ddcd912-32a0-43df-9908-63574f8c7e77" a
    LEFT JOIN RANKED r on a."property_id" = r."property_id"
WHERE
     "zip"::text = '02128'
    AND "owner"::text not in (
        'MASSACHUSETTS PORT AUTHORITY',
        'CITY OF BOSTON',
        'BOSTON HOUSING AUTHORITY',
        'ROMAN CATH ARCH OF BOS',
        'BROOKE CHARTER SCHOOL',
        'BROOKE CHARTER SCHOOL '
        'CITY OF BOSTON ',
        'MASSACHUSETTS PORT AUTH',
        'MASS PORT AUTHORITY',
        'CITY  OF  BOSTON',
        'BROOKE CHARTER SCHOOL '
    )
    AND "occupancytype"::text != 'Comm'
ORDER BY RANK ASC
limit 100;



And the link: https://nattaylor.com/labs/analyzeboston/#